## 1. Imports and initial transformations

In [1]:
# importing the libraries
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.impute import KNNImputer
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import KBinsDiscretizer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer, SimpleImputer
from sklearn.neighbors import LocalOutlierFactor
from matplotlib import pyplot as plt
import seaborn as sns
import re
import math
import datetime

In [2]:
# setting the options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
# pd.set_option('display.max_rows', None)
np.set_printoptions(threshold=np.inf)
plt.style.use('seaborn-v0_8-dark')

In [3]:
# importing the training and test data
df = pd.read_csv('../train_data.csv', sep=',')
df_test = pd.read_csv('../test_data.csv', sep=',')

C:\Users\joaom\AppData\Local\Temp\ipykernel_23016\672802336.py:2: DtypeWarning: Columns (29) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../train_data.csv', sep=',')


In [4]:
df.shape

(593471, 33)

In [5]:
# we check the first rows of the dataset
df.head()

Accident Date  Age at Injury Alternative Dispute Resolution Assembly Date  \
0    2019-12-30           31.0                              N    2020-01-01   
1    2019-08-30           46.0                              N    2020-01-01   
2    2019-12-06           40.0                              N    2020-01-01   
3           NaN            NaN                            NaN    2020-01-01   
4    2019-12-30           61.0                              N    2020-01-01   

  Attorney/Representative  Average Weekly Wage  Birth Year    C-2 Date  \
0                       N                 0.00      1988.0  2019-12-31   
1                       Y              1745.93      1973.0  2020-01-01   
2                       N              1434.80      1979.0  2020-01-01   
3                     NaN                  NaN         NaN         NaN   
4                       N                  NaN      1958.0  2019-12-31   

     C-3 Date                  Carrier Name Carrier Type  Claim Identifier  \
0         NaN    NEW HAMPSHIRE INSURANCE CO  1A. PRIVATE           5393875   
1  2020-01-14  ZURICH AMERICAN INSURANCE CO  1A. PRIVATE           5393091   
2         NaN     INDEMNITY INSURANCE CO OF  1A. PRIVATE           5393889   
3         NaN                           NaN          NaN         957648180   
4         NaN          STATE INSURANCE FUND      2A. SIF           5393887   

  Claim Injury Type County of Injury COVID-19 Indicator District Name  \
0       2. NON-COMP     ST. LAWRENCE                  N      SYRACUSE   
1      4. TEMPORARY          WYOMING                  N     ROCHESTER   
2      4. TEMPORARY           ORANGE                  N        ALBANY   
3               NaN              NaN                NaN           NaN   
4       2. NON-COMP         DUTCHESS                  N        ALBANY   

  First Hearing Date Gender  IME-4 Count  Industry Code  \
0                NaN      M          NaN           44.0   
1         2020-02-21      F          4.0           23.0   
2                NaN      M          NaN           56.0   
3                NaN    NaN          NaN            NaN   
4                NaN      M          NaN           62.0   

                                      Industry Code Description  \
0                                                  RETAIL TRADE   
1                                                  CONSTRUCTION   
2  ADMINISTRATIVE AND SUPPORT AND WASTE MANAGEMENT AND REMEDIAT   
3                                                           NaN   
4                             HEALTH CARE AND SOCIAL ASSISTANCE   

  Medical Fee Region  OIICS Nature of Injury Description  \
0                  I                                 NaN   
1                  I                                 NaN   
2                 II                                 NaN   
3                NaN                                 NaN   
4                 II                                 NaN   

   WCIO Cause of Injury Code WCIO Cause of Injury Description  \
0                       27.0     FROM LIQUID OR GREASE SPILLS   
1                       97.0                REPETITIVE MOTION   
2                       79.0   OBJECT BEING LIFTED OR HANDLED   
3                        NaN                              NaN   
4                       16.0  HAND TOOL, UTENSIL; NOT POWERED   

   WCIO Nature of Injury Code WCIO Nature of Injury Description  \
0                        10.0                         CONTUSION   
1                        49.0                    SPRAIN OR TEAR   
2                         7.0                        CONCUSSION   
3                         NaN                               NaN   
4                        43.0                          PUNCTURE   

   WCIO Part Of Body Code WCIO Part Of Body Description Zip Code  \
0                    62.0                      BUTTOCKS    13662   
1                    38.0                   SHOULDER(S)    14569   
2                    10.0          MULTIPLE HEAD INJURY  

In [6]:
# we check if there are any aggregation rows at the end of the dataset
df.tail()

Accident Date  Age at Injury Alternative Dispute Resolution  \
593466           NaN            NaN                            NaN   
593467    2022-12-13           72.0                              N   
593468           NaN            NaN                            NaN   
593469           NaN            NaN                            NaN   
593470           NaN            NaN                            NaN   

       Assembly Date Attorney/Representative  Average Weekly Wage  Birth Year  \
593466    2022-12-31                     NaN                  NaN         NaN   
593467    2022-12-31                       N                  0.0      1950.0   
593468    2022-12-31                     NaN                  NaN         NaN   
593469    2022-12-31                     NaN                  NaN         NaN   
593470    2022-12-31                     NaN                  NaN         NaN   

          C-2 Date C-3 Date                   Carrier Name Carrier Type  \
593466         NaN      NaN                            NaN          NaN   
593467  2022-12-31      NaN  TECHNOLOGY INSURANCE CO. INC.  1A. PRIVATE   
593468         NaN      NaN                            NaN          NaN   
593469         NaN      NaN                            NaN          NaN   
593470         NaN      NaN                            NaN          NaN   

        Claim Identifier Claim Injury Type County of Injury  \
593466         327160035               NaN              NaN   
593467           6165075       2. NON-COMP         SULLIVAN   
593468         249875936               NaN              NaN   
593469         120584215               NaN              NaN   
593470         818961390               NaN              NaN   

       COVID-19 Indicator District Name First Hearing Date Gender  \
593466                NaN           NaN                NaN    NaN   
593467                  N    BINGHAMTON                NaN      F   
593468                NaN           NaN                NaN    NaN   
593469                NaN           NaN                NaN    NaN   
593470                NaN           NaN                NaN    NaN   

        IME-4 Count  Industry Code       Industry Code Description  \
593466          NaN            NaN                             NaN   
593467          NaN           48.0  TRANSPORTATION AND WAREHOUSING   
593468          NaN            NaN                             NaN   
593469          NaN            NaN                             NaN   
593470          NaN            NaN                             NaN   

       Medical Fee Region  OIICS Nature of Injury Description  \
593466                NaN                                 NaN   
593467                  I                                 NaN   
593468                NaN                                 NaN   
593469                NaN                                 NaN   
593470                NaN                                 NaN   

        WCIO Cause of Injury Code  WCIO Cause of Injury Description  \
593466                        NaN                               NaN   
593467                       25.0  FROM DIFFERENT LEVEL (ELEVATION)   
593468                        NaN                               NaN   
593469                        NaN                               NaN   
593470                        NaN                               NaN   

        WCIO Nature of Injury Code WCIO Nature of Injury Description  \
593466                         NaN                               NaN   
593467                        90.0   MULTIPLE PHYSICAL INJURIES ONLY   
593468                         NaN                               NaN   
593469                         NaN                               NaN   
593470                         NaN                               NaN   

        WCIO Part Of Body Code WCIO Part Of Body Description Zip Code  \
593466                     NaN                           NaN      NaN   
593467                    -9.0      

### Index is now the Claim Identifier

In [7]:
df = df[~df['Claim Identifier'].duplicated()].set_index('Claim Identifier').rename_axis(None)

In [8]:
df.head()

Accident Date  Age at Injury Alternative Dispute Resolution  \
5393875      2019-12-30           31.0                              N   
5393091      2019-08-30           46.0                              N   
5393889      2019-12-06           40.0                              N   
957648180           NaN            NaN                            NaN   
5393887      2019-12-30           61.0                              N   

          Assembly Date Attorney/Representative  Average Weekly Wage  \
5393875      2020-01-01                       N                 0.00   
5393091      2020-01-01                       Y              1745.93   
5393889      2020-01-01                       N              1434.80   
957648180    2020-01-01                     NaN                  NaN   
5393887      2020-01-01                       N                  NaN   

           Birth Year    C-2 Date    C-3 Date                  Carrier Name  \
5393875        1988.0  2019-12-31         NaN    NEW HAMPSHIRE INSURANCE CO   
5393091        1973.0  2020-01-01  2020-01-14  ZURICH AMERICAN INSURANCE CO   
5393889        1979.0  2020-01-01         NaN     INDEMNITY INSURANCE CO OF   
957648180         NaN         NaN         NaN                           NaN   
5393887        1958.0  2019-12-31         NaN          STATE INSURANCE FUND   

          Carrier Type Claim Injury Type County of Injury COVID-19 Indicator  \
5393875    1A. PRIVATE       2. NON-COMP     ST. LAWRENCE                  N   
5393091    1A. PRIVATE      4. TEMPORARY          WYOMING                  N   
5393889    1A. PRIVATE      4. TEMPORARY           ORANGE                  N   
957648180          NaN               NaN              NaN                NaN   
5393887        2A. SIF       2. NON-COMP         DUTCHESS                  N   

          District Name First Hearing Date Gender  IME-4 Count  Industry Code  \
5393875        SYRACUSE                NaN      M          NaN           44.0   
5393091       ROCHESTER         2020-02-21      F          4.0           23.0   
5393889          ALBANY                NaN      M          NaN           56.0   
957648180           NaN                NaN    NaN          NaN            NaN   
5393887          ALBANY                NaN      M          NaN           62.0   

                                              Industry Code Description  \
5393875                                                    RETAIL TRADE   
5393091                                                    CONSTRUCTION   
5393889    ADMINISTRATIVE AND SUPPORT AND WASTE MANAGEMENT AND REMEDIAT   
957648180                                                           NaN   
5393887                               HEALTH CARE AND SOCIAL ASSISTANCE   

          Medical Fee Region  OIICS Nature of Injury Description  \
5393875                    I                                 NaN   
5393091                    I                                 NaN   
5393889                   II                                 NaN   
957648180                NaN                                 NaN   
5393887                   II                                 NaN   

           WCIO Cause of Injury Code WCIO Cause of Injury Description  \
5393875                         27.0     FROM LIQUID OR GREASE SPILLS   
5393091                         97.0                REPETITIVE MOTION   
5393889                         79.0   OBJECT BEING LIFTED OR HANDLED   
957648180                        NaN                              NaN   
5393887                         16.0  HAND TOOL, UTENSIL; NOT POWERED   

           WCIO Nature of Injury Code WCIO Nature of Injury Description  \
5393875                          10.0                         CONTUSION   
5393091                          49.0                    SPRAIN OR TEAR   
5393889                           7.0                        CONCUSSION   
957648180                         NaN                               NaN   
5393887              

In [9]:
# we check the datatypes and null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 593470 entries, 5393875 to 818961390
Data columns (total 32 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   Accident Date                       570337 non-null  object 
 1   Age at Injury                       574026 non-null  float64
 2   Alternative Dispute Resolution      574026 non-null  object 
 3   Assembly Date                       593470 non-null  object 
 4   Attorney/Representative             574026 non-null  object 
 5   Average Weekly Wage                 545375 non-null  float64
 6   Birth Year                          544948 non-null  float64
 7   C-2 Date                            559466 non-null  object 
 8   C-3 Date                            187245 non-null  object 
 9   Carrier Name                        574026 non-null  object 
 10  Carrier Type                        574026 non-null  object 
 11  Claim Injury Type     

In [10]:
# Changing the data type of the values to string
df['Zip Code'] = df['Zip Code'].astype(str)

**Data type analysis:**

Features that should be integers:
- 'Age at Injury'
- 'Birth Year'
- 'IME-4 Count'
- 'Industry Code'
- 'WCIO Cause of Injury Code'
- 'WCIO Nature of Injury Code'
- 'WCIO Part Of Body Code'
- 'Number of Dependents'

Features that should be booleans:
- 'Agreement Reached'

In [11]:
features_to_cast = ['Age at Injury', 'Birth Year', 'IME-4 Count', 'Industry Code', 'WCIO Cause of Injury Code', 'WCIO Nature of Injury Code', 'WCIO Part Of Body Code', 'Number of Dependents', 'Agreement Reached']

In [12]:
for feature in features_to_cast:
    df[feature] = pd.to_numeric(df[feature], errors='coerce').astype('Int64')

In [13]:
date_columns = ['Accident Date', 'Assembly Date', 'C-2 Date', 'C-3 Date', 'First Hearing Date']

for col in date_columns:
    # first, we convert the columns to datetime
    df[col] = pd.to_datetime(df[col])
    # then we convert it to int
    # df[col] = df[col].dt.strftime('%Y%m%d').astype('Int64')

In [14]:
df['Zip Code'] = df['Zip Code'].replace('nan', np.nan)

In [15]:
df.drop_duplicates(inplace=True)

In [16]:
df = df[~df['Claim Injury Type'].isna()]

## 2. Data Preprocessing

### 2.1. Weird values

As we previously mentioned, there are some columns with weird values:
- **Age at Injury**: multiple values below 14, which is the minimum legal age to work in the USA
- **Birth Year**: multiple 0 values
- **WCIO Part Of Body Code**: has a negative value

Before we go any further, let's try to tackle these issues.

#### 2.1.1. Age at Injury

In [17]:
len(df[df['Age at Injury'] < 14])

5510

In [18]:
df[df['Age at Injury'] < 14].head()

Accident Date  Age at Injury Alternative Dispute Resolution  \
5393971    2019-06-26              0                              N   
5394093    2019-12-28              0                              N   
5394068           NaT              0                              N   
5393794           NaT              0                              N   
5393876    2019-12-20              0                              N   

        Assembly Date Attorney/Representative  Average Weekly Wage  \
5393971    2020-01-02                       N                  0.0   
5394093    2020-01-02                       N                  0.0   
5394068    2020-01-02                       Y                  0.0   
5393794    2020-01-02                       Y                  0.0   
5393876    2020-01-02                       N                  0.0   

         Birth Year   C-2 Date   C-3 Date                    Carrier Name  \
5393971           0        NaT        NaT    AMERICAN ZURICH INSURANCE CO   
5394093           0 2019-12-30        NaT          TWIN CITY FIRE INS CO.   
5394068        1959        NaT 2019-12-23  STARR INDEMNITY & LIABILITY CO   
5393794        1984        NaT 2019-12-27               LM INSURANCE CORP   
5393876           0 2019-12-30        NaT  CARMEL CENTRAL SCHOOL DISTRICT   

            Carrier Type Claim Injury Type County of Injury  \
5393971      1A. PRIVATE       2. NON-COMP           NASSAU   
5394093      1A. PRIVATE       2. NON-COMP         NEW YORK   
5394068      1A. PRIVATE       2. NON-COMP            BRONX   
5393794      1A. PRIVATE       2. NON-COMP         ONONDAGA   
5393876  3A. SELF PUBLIC       2. NON-COMP           PUTNAM   

        COVID-19 Indicator District Name First Hearing Date Gender  \
5393971                  N           NYC                NaT      M   
5394093                  N           NYC                NaT      M   
5394068                  N           NYC                NaT      M   
5393794                  N      SYRACUSE                NaT      F   
5393876                  N        ALBANY                NaT      F   

         IME-4 Count  Industry Code            Industry Code Description  \
5393971         <NA>             62    HEALTH CARE AND SOCIAL ASSISTANCE   
5394093         <NA>             71  ARTS, ENTERTAINMENT, AND RECREATION   
5394068         <NA>             23                         CONSTRUCTION   
5393794         <NA>           <NA>                                  NaN   
5393876         <NA>           <NA>                                  NaN   

        Medical Fee Region  OIICS Nature of Injury Description  \
5393971                 IV                                 NaN   
5394093                 IV                                 NaN   
5394068                 IV                                 NaN   
5393794                 II                                 NaN   
5393876                III                                 NaN   

         WCIO Cause of Injury Code        WCIO Cause of Injury Description  \
5393971                       <NA>                                     NaN   
5394093                         55                     HOLDING OR CARRYING   
5394068                       <NA>                                     NaN   
5393794                       <NA>                                     NaN   
5393876                         74  FELLOW WORKER, PATIENT OR OTHER PERSON   

         WCIO Nature of Injury Code WCIO Nature of Injury Description  \
5393971                        <NA>                               NaN   
5394093                          52                    STRAIN OR TEAR   
5394068                        <NA>                               NaN   
5393794                        <NA>                               NaN   
5393876                          40                        LACERATION   

         WCIO Part Of Body Code WCIO Part Of Body Description Zip Code  \
5393971                    <NA>                           NaN

In [19]:
# This code is used to determine the number of individuals in the dataset who were under 14 years old at the time of their injury, 
# whose birth year is recorded and not given as 0, and who have a recorded (non-missing) accident date and birth year. 
len(df[(df['Age at Injury'] < 14) & (df['Birth Year'] != 0) & ~(df['Accident Date'].isna()) & ~(df['Birth Year'].isna())])

122

We can manually calculate these 122 values, since we have information about 'Birth Year' and 'Accident Date'

In [20]:
df[(df['Age at Injury'] < 14) & (df['Birth Year'] != 0) & ~(df['Accident Date'].isna()) & ~(df['Birth Year'].isna())].head()

Accident Date  Age at Injury Alternative Dispute Resolution  \
5399016    2019-08-11              0                              N   
5401957    2020-01-12              0                              N   
5406866    2020-01-06              0                              N   
5412625    2020-01-09              1                              N   
5412736    2020-01-09              1                              N   

        Assembly Date Attorney/Representative  Average Weekly Wage  \
5399016    2020-01-09                       N               150.00   
5401957    2020-01-14                       N               528.71   
5406866    2020-01-17                       N                 0.00   
5412625    2020-01-25                       N                 0.00   
5412736    2020-01-25                       N                  NaN   

         Birth Year   C-2 Date   C-3 Date                  Carrier Name  \
5399016        1901 2020-01-09 2019-12-27       WESCO INSURANCE COMPANY   
5401957        1901 2020-01-14        NaT  AMTRUST INS CO OF KANSAS INC   
5406866        1901 2020-01-17        NaT   INDEMNITY INS. OF N AMERICA   
5412625        2018 2020-01-24        NaT    NEW HAMPSHIRE INSURANCE CO   
5412736        2018 2020-01-24        NaT    NEW HAMPSHIRE INSURANCE CO   

        Carrier Type Claim Injury Type County of Injury COVID-19 Indicator  \
5399016  1A. PRIVATE       3. MED ONLY            KINGS                  N   
5401957  1A. PRIVATE      4. TEMPORARY     ST. LAWRENCE                  N   
5406866  1A. PRIVATE       2. NON-COMP             ERIE                  N   
5412625  1A. PRIVATE       2. NON-COMP          SUFFOLK                  N   
5412736  1A. PRIVATE       2. NON-COMP          SUFFOLK                  N   

        District Name First Hearing Date Gender  IME-4 Count  Industry Code  \
5399016           NYC                NaT      M         <NA>             72   
5401957        ALBANY                NaT      M         <NA>             31   
5406866       BUFFALO                NaT      F         <NA>             54   
5412625     HAUPPAUGE                NaT      F         <NA>             54   
5412736     HAUPPAUGE                NaT      F         <NA>             54   

                                Industry Code Description Medical Fee Region  \
5399016                   ACCOMMODATION AND FOOD SERVICES                 IV   
5401957                                     MANUFACTURING                  I   
5406866  PROFESSIONAL, SCIENTIFIC, AND TECHNICAL SERVICES                  I   
5412625  PROFESSIONAL, SCIENTIFIC, AND TECHNICAL SERVICES                 IV   
5412736  PROFESSIONAL, SCIENTIFIC, AND TECHNICAL SERVICES                 IV   

         OIICS Nature of Injury Description  WCIO Cause of Injury Code  \
5399016                                 NaN                         45   
5401957                                 NaN                         19   
5406866                                 NaN                         31   
5412625                                 NaN                         31   
5412736                                 NaN                         31   

                    WCIO Cause of Injury Description  \
5399016  COLLISION OR SIDESWIPE WITH ANOTHER VEHICLE   
5401957                   CUT, PUNCTURE, SCRAPE, NOC   
5406866                      FALL, SLIP OR TRIP, NOC   
5412625                      FALL, SLIP OR TRIP, NOC   
5412736                      FALL, SLIP OR TRIP, NOC   

         WCIO Nature of Injury Code WCIO Nature of Injury Description  \
5399016                          90   MULTIPLE PHYSICAL INJURIES ONLY   
5401957                          40                        LACERATION   
5406866                          59  ALL OTHER SPECIFIC INJURIES, NOC   
5412625                          10                         CONTUSION   
5412736                          10                         CONTUSION   

         WCIO Part Of Body Code        WCIO Part Of Body Descri

In [21]:
df['Age at Injury'] = np.where(
    # we select only the rows that meet the necessary conditions
    (df['Age at Injury'] < 14) & (df['Birth Year'] != 0) & (df['Accident Date'].notna()) & (df['Birth Year'].notna()),
    # if conditions are met, we calculate the new value
    df['Accident Date'].dt.year - df['Birth Year'],
    # otherwise, we keep the original value
    df['Age at Injury']
)

In [22]:
len(df[(df['Age at Injury'] < 14) & (df['Birth Year'] != 0) & (df['Accident Date'].notna()) & (df['Birth Year'].notna())])

27

Apparently we still have some inconsistent data, but we were able to correct around 100 rows.

#### 2.1.2. Birth Year

In [23]:
len(df[df['Birth Year'] == 0])

25081

In [24]:
len(df[(df['Birth Year'] == 0) & (df['Accident Date'].notna()) & (df['Age at Injury'] >= 14)])

23141

In [25]:
df[(df['Birth Year'] == 0) & ~(df['Accident Date'].notna()) & (df['Age at Injury'] >= 14)].head()

Empty DataFrame
Columns: [Accident Date, Age at Injury, Alternative Dispute Resolution, Assembly Date, Attorney/Representative, Average Weekly Wage, Birth Year, C-2 Date, C-3 Date, Carrier Name, Carrier Type, Claim Injury Type, County of Injury, COVID-19 Indicator, District Name, First Hearing Date, Gender, IME-4 Count, Industry Code, Industry Code Description, Medical Fee Region, OIICS Nature of Injury Description, WCIO Cause of Injury Code, WCIO Cause of Injury Description, WCIO Nature of Injury Code, WCIO Nature of Injury Description, WCIO Part Of Body Code, WCIO Part Of Body Description, Zip Code, Agreement Reached, WCB Decision, Number of Dependents]
Index: []

We can manually impute most of the weird values just by doing some simple math.

In [26]:
df['Birth Year'] = np.where(
    # we select only the rows that meet the necessary conditions
    (df['Birth Year'] == 0) & (df['Birth Year'].notna()) & (df['Accident Date'].notna()) & (df['Age at Injury'] >= 14),
    # if conditions are met, we calculate the new value
    df['Accident Date'].dt.year - df['Age at Injury'],
    # otherwise, we keep the original value
    df['Birth Year']
)

In [27]:
len(df[df['Birth Year'] == 0])

1940

We were able to significantly reduce the amount of weird values

#### 2.1.3. WCIO Part Of Body Code

In [28]:
len(df[df['WCIO Part Of Body Code'] < 0])

42011

In [29]:
df[df['WCIO Part Of Body Code'] < 0]['WCIO Part Of Body Code'].unique()

<IntegerArray>
[-9]
Length: 1, dtype: Int64

All negative values are the same value - let's check if there are any values '9' or if we can simply convert these values to its absolute value.

In [30]:
len(df[df['WCIO Part Of Body Code'] == 9])

0

As there are no values that take the value '9', we will convert the negative values to the absolute value

In [31]:
df['WCIO Part Of Body Code'] = np.where(
    # we select only the rows that meet the necessary conditions
    (df['WCIO Part Of Body Code'] < 0) & (df['WCIO Part Of Body Code'].notna()),
    # if conditions are met, we calculate the new value
    df['WCIO Part Of Body Code'].abs(),
    # otherwise, we keep the original value
    df['WCIO Part Of Body Code']
)

In [32]:
len(df[df['WCIO Part Of Body Code'] < 0])

0

#### 2.1.4. Datetime variables

Some dates happen before the accident happened

In [33]:
# Check that the DataFrame has a multi-index with the keys as one of the levels
# If not sure, you can re-create the DataFrame with a clear multi-index as follows:
invalid_orders_df = pd.concat([
    df[df['Assembly Date'] < df['Accident Date']],
    df[df['C-2 Date'] < df['Accident Date']],
    df[df['C-3 Date'] < df['Accident Date']],
    df[df['First Hearing Date'] < df['Accident Date']]
], keys=[
    'Assembly Date before Accident Date',
    'C-2 Date before Accident Date',
    'C-3 Date before Accident Date',
    'First Hearing Date before Accident Date'
], names=['Date Issue', 'Index'])

# Specify the datetime columns
datetime_columns = [
    'Accident Date',
    'Assembly Date',
    'C-2 Date',
    'C-3 Date',
    'First Hearing Date'
]


# Display the head of the DataFrame for the datetime columns
invalid_orders_df[datetime_columns].head()

Accident Date Assembly Date  \
Date Issue                         Index                                 
Assembly Date before Accident Date 5393953    2020-02-26    2020-01-01   
                                   5394099    2020-01-23    2020-01-02   
                                   5393830    2020-03-01    2020-01-02   
                                   5394613    2020-01-27    2020-01-03   
                                   5394770    2020-01-23    2020-01-03   

                                             C-2 Date   C-3 Date  \
Date Issue                         Index                           
Assembly Date before Accident Date 5393953 2019-12-31 2020-01-31   
                                   5394099 2020-01-14 2019-12-27   
                                   5393830 2020-01-17 2019-12-27   
                                   5394613 2020-01-15 2019-12-30   
                                   5394770 2020-01-06 2019-12-26   

                                           First Hearing Date  
Date Issue                         Index                       
Assembly Date before Accident Date 5393953         2020-05-04  
                                   5394099         2020-06-05  
                                   5393830         2020-08-28  
                                   5394613         2021-09-01  
                                   5394770         2020-05-29

In [34]:
# Now calculate value_counts for 'Claim Injury Type' for each key
# This iterates over each key in the multi-index and performs value_counts on 'Claim Injury Type'
for key in invalid_orders_df.index.levels[0]:  # Loop through each key
    print(f"Counts for {key}:")
    # Extract the subset for the current key and perform value_counts on 'Claim Injury Type'
    claim_type_counts = invalid_orders_df.xs(key, level='Date Issue')['Claim Injury Type'].value_counts()
    print(claim_type_counts)
    print("\n")

Counts for Assembly Date before Accident Date:
Claim Injury Type
5. PPD SCH LOSS    723
3. MED ONLY        350
4. TEMPORARY       255
2. NON-COMP         54
6. PPD NSL          14
1. CANCELLED         7
7. PTD               3
8. DEATH             1
Name: count, dtype: int64


Counts for C-2 Date before Accident Date:
Claim Injury Type
5. PPD SCH LOSS    404
4. TEMPORARY       256
3. MED ONLY        252
2. NON-COMP         48
6. PPD NSL          12
1. CANCELLED         7
7. PTD               2
8. DEATH             1
Name: count, dtype: int64


Counts for C-3 Date before Accident Date:
Claim Injury Type
5. PPD SCH LOSS    700
3. MED ONLY        314
4. TEMPORARY       192
2. NON-COMP         58
6. PPD NSL          14
1. CANCELLED         9
7. PTD               2
Name: count, dtype: int64


Counts for First Hearing Date before Accident Date:
Claim Injury Type
3. MED ONLY        34
5. PPD SCH LOSS    18
4. TEMPORARY       15
2. NON-COMP         5
6. PPD NSL          2
Name: count, dtype: in

Maybe this discrepancy can affect the target variable

In [35]:
columns_to_display = datetime_columns + ['Claim Injury Type']

invalid_orders_df[columns_to_display].head(5)

Accident Date Assembly Date  \
Date Issue                         Index                                 
Assembly Date before Accident Date 5393953    2020-02-26    2020-01-01   
                                   5394099    2020-01-23    2020-01-02   
                                   5393830    2020-03-01    2020-01-02   
                                   5394613    2020-01-27    2020-01-03   
                                   5394770    2020-01-23    2020-01-03   

                                             C-2 Date   C-3 Date  \
Date Issue                         Index                           
Assembly Date before Accident Date 5393953 2019-12-31 2020-01-31   
                                   5394099 2020-01-14 2019-12-27   
                                   5393830 2020-01-17 2019-12-27   
                                   5394613 2020-01-15 2019-12-30   
                                   5394770 2020-01-06 2019-12-26   

                                           First Hearing Date  \
Date Issue                         Index                        
Assembly Date before Accident Date 5393953         2020-05-04   
                                   5394099         2020-06-05   
                                   5393830         2020-08-28   
                                   5394613         2021-09-01   
                                   5394770         2020-05-29   

                                           Claim Injury Type  
Date Issue                         Index                      
Assembly Date before Accident Date 5393953   5. PPD SCH LOSS  
                                   5394099        6. PPD NSL  
                                   5393830   5. PPD SCH LOSS  
                                   5394613   5. PPD SCH LOSS  
                                   5394770   5. PPD SCH LOSS

In [36]:
invalid_orders_df['Claim Injury Type'].value_counts().sort_values()

Claim Injury Type
8. DEATH              2
7. PTD                7
1. CANCELLED         23
6. PPD NSL           42
2. NON-COMP         165
4. TEMPORARY        718
3. MED ONLY         950
5. PPD SCH LOSS    1845
Name: count, dtype: int64

In [37]:
# List all columns, then remove those you want to exclude from checking duplicates
columns_to_check = invalid_orders_df.columns.difference([
    'Accident Date', 'Assembly Date', 'C-2 Date', 'C-3 Date', 'First Hearing Date'
])

# Check for duplicates based on the remaining columns
duplicates = invalid_orders_df.duplicated(subset=columns_to_check, keep=False)

# Count duplicates
duplicate_count = duplicates.sum()


print(f"Number of different rows (excluding specified date columns): {duplicate_count}")

# Optional: Display rows that are duplicates
# duplicate_rows = invalid_orders_df[duplicates]
# print(duplicate_rows)

Number of different rows (excluding specified date columns): 3453


This discrepancy happens 3453 times

### 2.2. Missing values

In [38]:
# Calculate the number of NaNs for each column
nan_counts = df.isna().sum()

# Get the total number of rows (entries) in the DataFrame
total_rows = df.shape[0]

# Calculate the percentage of NaN values for each column
percentage_nans = (nan_counts / total_rows) * 100

# Format the percentage with '%' sign
percentage_nans = percentage_nans.apply(lambda x: f"{x:.2f}%")

# Combine all information into a DataFrame for better readability
nan_summary = pd.DataFrame({
    'NaN Count': nan_counts,
    'Total Values': [total_rows] * len(nan_counts),  # Ensure this column matches the length of nan_counts
    'Percentage NaN': percentage_nans
})

# Print the result
print("Summary of NaN values per column:\n")
print(nan_summary)

Summary of NaN values per column:

                                    NaN Count  Total Values Percentage NaN
Accident Date                            3688        574025          0.64%
Age at Injury                               0        574025          0.00%
Alternative Dispute Resolution              0        574025          0.00%
Assembly Date                               0        574025          0.00%
Attorney/Representative                     0        574025          0.00%
Average Weekly Wage                     28651        574025          4.99%
Birth Year                              29078        574025          5.07%
C-2 Date                                14559        574025          2.54%
C-3 Date                               386781        574025         67.38%
Carrier Name                                0        574025          0.00%
Carrier Type                                0        574025          0.00%
Claim Injury Type                           0        574025      

In [39]:
missing_values = [
    'Accident Date'
    , 'Average Weekly Wage'
    , 'Birth Year'
    #, 'C-2 Date'  # missing form could have relationship with the target
    #, 'C-3 Date'  # missing form could have relationship with the target
    #, 'First Hearing Date'  # missing values means no hearing has held
    #, 'IME-4 Count'  # missing form could have relationship with the target
    , 'Industry Code'
    #, 'Industry Code Description'  # we will only use the numeric form of this variable
    #, 'OIICS Nature of Injury Description'  # only missing values, so we will not use this feature
    , 'WCIO Cause of Injury Code'
    #, 'WCIO Cause of Injury Description'  # we will only use the numeric form of this variable
    , 'WCIO Nature of Injury Code'
    #, 'WCIO Nature of Injury Description'  # we will only use the numeric form of this variable
    , 'WCIO Part Of Body Code'
    #, 'WCIO Part Of Body Description'  # we will only use the numeric form of this variable
    , 'Zip Code'
]

We will not impute missing values in the commented columns, as per the explainations in the comments

### 2.3. Categorical Encoding

In [40]:
# we can drop the descriptive columns and the unary column ('WCB Decision')
df.drop(columns=['WCIO Part Of Body Description', 'Industry Code Description', 'WCIO Nature of Injury Description', 'WCIO Cause of Injury Description', 'OIICS Nature of Injury Description', 'WCB Decision'], inplace=True)

In [41]:
categorical_columns = [
    'Alternative Dispute Resolution'
    , 'Attorney/Representative'
    , 'Carrier Type'
    , 'Carrier Name'
    , 'Claim Injury Type'
    , 'County of Injury'
    , 'COVID-19 Indicator'
    , 'District Name'
    , 'Gender'
    , 'Medical Fee Region'
    , 'Zip Code'
]

data = df[categorical_columns]

data = data.dropna()

# Display the first few rows to verify the remaining columns
data.head()


Alternative Dispute Resolution Attorney/Representative Carrier Type  \
5393875                              N                       N  1A. PRIVATE   
5393091                              N                       Y  1A. PRIVATE   
5393889                              N                       N  1A. PRIVATE   
5393887                              N                       N      2A. SIF   
5393863                              N                       N  1A. PRIVATE   

                         Carrier Name Claim Injury Type County of Injury  \
5393875    NEW HAMPSHIRE INSURANCE CO       2. NON-COMP     ST. LAWRENCE   
5393091  ZURICH AMERICAN INSURANCE CO      4. TEMPORARY          WYOMING   
5393889     INDEMNITY INSURANCE CO OF      4. TEMPORARY           ORANGE   
5393887          STATE INSURANCE FUND       2. NON-COMP         DUTCHESS   
5393863   INDEMNITY INS. OF N AMERICA       3. MED ONLY          SUFFOLK   

        COVID-19 Indicator District Name Gender Medical Fee Region Zip Code  
5393875                  N      SYRACUSE      M                  I    13662  
5393091                  N     ROCHESTER      F                  I    14569  
5393889                  N        ALBANY      M                 II    12589  
5393887                  N        ALBANY      M                 II    12603  
5393863                  N     HAUPPAUGE      M                 IV    11772

In [42]:
ordinal_encoder = OrdinalEncoder()

In [43]:
data = df.copy()
data[categorical_columns] = ordinal_encoder.fit_transform(df[categorical_columns])

In [44]:
def to_ordinal(data, column):
    data[column] = data[column].apply(lambda x: x.toordinal() if pd.notnull(x) else np.nan).astype('Int64')

In [45]:
data['Accident Date'] = pd.to_datetime(df['Accident Date'])
# df['Assembly Date'] = pd.to_datetime(df['Assembly Date'])

In [46]:
dates = [
    'Accident Date'
    , 'Assembly Date'
    , 'C-2 Date'
    , 'C-3 Date'
    , 'First Hearing Date'
]

In [47]:
data.dtypes

Accident Date                     datetime64[ns]
Age at Injury                            float64
Alternative Dispute Resolution           float64
Assembly Date                     datetime64[ns]
Attorney/Representative                  float64
Average Weekly Wage                      float64
Birth Year                               float64
C-2 Date                          datetime64[ns]
C-3 Date                          datetime64[ns]
Carrier Name                             float64
Carrier Type                             float64
Claim Injury Type                        float64
County of Injury                         float64
COVID-19 Indicator                       float64
District Name                            float64
First Hearing Date                datetime64[ns]
Gender                                   float64
IME-4 Count                                Int64
Industry Code                              Int64
Medical Fee Region                       float64
WCIO Cause of Injury

In [48]:
for col in dates:
    to_ordinal(data, col)

In [49]:
data.dtypes

Accident Date                       Int64
Age at Injury                     float64
Alternative Dispute Resolution    float64
Assembly Date                       Int64
Attorney/Representative           float64
Average Weekly Wage               float64
Birth Year                        float64
C-2 Date                            Int64
C-3 Date                            Int64
Carrier Name                      float64
Carrier Type                      float64
Claim Injury Type                 float64
County of Injury                  float64
COVID-19 Indicator                float64
District Name                     float64
First Hearing Date                  Int64
Gender                            float64
IME-4 Count                         Int64
Industry Code                       Int64
Medical Fee Region                float64
WCIO Cause of Injury Code           Int64
WCIO Nature of Injury Code          Int64
WCIO Part Of Body Code            float64
Zip Code                          

All columns are now numeric

In [50]:
data.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875         737423           31.0                             0.0   
5393091         737301           46.0                             0.0   
5393889         737399           40.0                             0.0   
5393887         737423           61.0                             0.0   
5393863         737419           67.0                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875         737425                      0.0                 0.00   
5393091         737425                      1.0              1745.93   
5393889         737425                      0.0              1434.80   
5393887         737425                      0.0                  NaN   
5393863         737425                      0.0                 0.00   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875      1988.0    737424      <NA>        1197.0           0.0   
5393091      1973.0    737425    737438        2044.0           0.0   
5393889      1979.0    737425      <NA>         894.0           0.0   
5393887      1958.0    737424      <NA>        1710.0           1.0   
5393863      1952.0    737424      <NA>         893.0           0.0   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875                1.0              49.0                 0.0   
5393091                3.0              61.0                 0.0   
5393889                3.0              35.0                 0.0   
5393887                1.0              13.0                 0.0   
5393863                2.0              51.0                 0.0   

         District Name  First Hearing Date  Gender  IME-4 Count  \
5393875            7.0                <NA>     1.0         <NA>   
5393091            5.0              737476     0.0            4   
5393889            0.0                <NA>     1.0         <NA>   
5393887            0.0                <NA>     1.0         <NA>   
5393863            3.0                <NA>     1.0         <NA>   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875             44                 0.0                         27   
5393091             23                 0.0                         97   
5393889             56                 1.0                         79   
5393887             62                 1.0                         16   
5393863             44                 3.0                         31   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                          10                    62.0    3935.0   
5393091                          49                    38.0    4606.0   
5393889                           7                    10.0    3075.0   
5393887                          43                    36.0    3088.0   
5393863                          10                    38.0    2362.0   

         Agreement Reached  Number of Dependents  
5393875                  0                     1  
5393091                  1                     4  
5393889                  0                     6  
5393887                  0                     1  
5393863                  0                     5

For the features where missing values carry meaning, one could leave them as NaN and try an algorithm that can take advantage and learn from these missing values. We will analyse the missing values impact on the target later in this section.</br></br>
For the remaining columns, we will impute these values, using KNN Imputer. For that, we will take a small sample of our data, without missing values, and test the imputer to make a choice regarding the optimal number of neighbors.

### 2.4. Outliers

In [51]:
lof = LocalOutlierFactor(n_neighbors=20, n_jobs=-1)

In [52]:
test = df.sample(n=100000, random_state=17)

In [53]:
# initialize the imputer
imputer = SimpleImputer(strategy='median')
# create a new df with no missing values
df_temp_imputed = pd.DataFrame(imputer.fit_transform(data), columns=data.columns)

# Run LOF
outlier_labels = lof.fit_predict(df_temp_imputed)

# Remove rows marked as outliers
data_no_outliers = data[outlier_labels != -1]

In [54]:
data_no_outliers.shape

(559202, 26)

In [55]:
# % lines removed
(1- (data_no_outliers.shape[0] / df.shape[0])) * 100

2.5822917120334465

### 2.5. Feature Scaling

In [56]:
min_max_scaler = MinMaxScaler()

In [57]:
df_scaled = min_max_scaler.fit_transform(data_no_outliers)

In [58]:
df_scaled = pd.DataFrame(df_scaled, columns = data_no_outliers.columns).set_index(data_no_outliers.index)

In [59]:
len(data_no_outliers['Accident Date'].unique())

5405

In [60]:
df_scaled.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875       0.939606       0.258333                             0.0   
5393091       0.934224       0.383333                             0.0   
5393889       0.938548       0.333333                             0.0   
5393887       0.939606       0.508333                             0.0   
5393863       0.939430       0.558333                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875            0.0                      0.0             0.000000   
5393091            0.0                      1.0             0.062790   
5393889            0.0                      0.0             0.051601   
5393887            0.0                      0.0                  NaN   
5393863            0.0                      0.0             0.000000   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875    0.985134  0.848420       NaN      0.585330      0.000000   
5393091    0.977701  0.848517    0.1846      0.999511      0.000000   
5393889    0.980674  0.848517       NaN      0.437164      0.000000   
5393887    0.970268  0.848420       NaN      0.836186      0.142857   
5393863    0.967294  0.848420       NaN      0.436675      0.000000   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875           0.142857          0.790323                 0.0   
5393091           0.428571          0.983871                 0.0   
5393889           0.428571          0.564516                 0.0   
5393887           0.142857          0.209677                 0.0   
5393863           0.285714          0.822581                 0.0   

         District Name  First Hearing Date    Gender  IME-4 Count  \
5393875       1.000000                 NaN  0.333333          NaN   
5393091       0.714286            0.013836  0.000000     0.041667   
5393889       0.000000                 NaN  0.333333          NaN   
5393887       0.000000                 NaN  0.333333          NaN   
5393863       0.428571                 NaN  0.333333          NaN   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875       0.407407                0.00                   0.265306   
5393091       0.148148                0.00                   0.979592   
5393889       0.555556                0.25                   0.795918   
5393887       0.629630                0.25                   0.153061   
5393863       0.407407                0.75                   0.306122   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                    0.100000                0.588889  0.391192   
5393091                    0.533333                0.322222  0.457898   
5393889                    0.066667                0.011111  0.305696   
5393887                    0.466667                0.300000  0.306989   
5393863                    0.100000                0.322222  0.234815   

         Agreement Reached  Number of Dependents  
5393875                0.0              0.166667  
5393091                1.0              0.666667  
5393889                0.0              1.000000  
5393887                0.0              0.166667  
5393863                0.0              0.833333

In [ ]:
#df_scaled.to_csv('train_data_scaled.csv', index=True)

### 2.6. Data Imputation

In [ ]:
# numeric_feats = [
#     'Accident Date'
#     , 'Age at Injury'
#     , 'Assembly Date'
#     , 'Average Weekly Wage'
#     , 'Birth Year'
#     , 'C-2 Date'
#     , 'C-3 Date'
#     , 'First Hearing Date'
#     , 'IME-4 Count'
#     , 'Number of Dependents'
# ]

# cat_feats = [
#     'Alternative Dispute Resolution'
#     , 'Attorney/Representative'
#     , 'Carrier Name'
#     , 'Carrier Type'
#     , 'Claim Injury Type'
#     , 'County of Injury'
#     , 'COVID-19 Indicator'
#     , 'District Name'
#     , 'Gender'
#     , 'Industry Code'
#     , 'Medical Fee Region'
#     , 'WCIO Cause of Injury Code'
#     , 'WCIO Nature of Injury Code'
#     , 'WCIO Part Of Body Code'
#     , 'Zip Code'
#     , 'Agreement Reached'
# ]

In [ ]:
#df_scaled = pd.read_csv('train_data_scaled.csv', sep=',')

In [61]:
df_scaled.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875       0.939606       0.258333                             0.0   
5393091       0.934224       0.383333                             0.0   
5393889       0.938548       0.333333                             0.0   
5393887       0.939606       0.508333                             0.0   
5393863       0.939430       0.558333                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875            0.0                      0.0             0.000000   
5393091            0.0                      1.0             0.062790   
5393889            0.0                      0.0             0.051601   
5393887            0.0                      0.0                  NaN   
5393863            0.0                      0.0             0.000000   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875    0.985134  0.848420       NaN      0.585330      0.000000   
5393091    0.977701  0.848517    0.1846      0.999511      0.000000   
5393889    0.980674  0.848517       NaN      0.437164      0.000000   
5393887    0.970268  0.848420       NaN      0.836186      0.142857   
5393863    0.967294  0.848420       NaN      0.436675      0.000000   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875           0.142857          0.790323                 0.0   
5393091           0.428571          0.983871                 0.0   
5393889           0.428571          0.564516                 0.0   
5393887           0.142857          0.209677                 0.0   
5393863           0.285714          0.822581                 0.0   

         District Name  First Hearing Date    Gender  IME-4 Count  \
5393875       1.000000                 NaN  0.333333          NaN   
5393091       0.714286            0.013836  0.000000     0.041667   
5393889       0.000000                 NaN  0.333333          NaN   
5393887       0.000000                 NaN  0.333333          NaN   
5393863       0.428571                 NaN  0.333333          NaN   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875       0.407407                0.00                   0.265306   
5393091       0.148148                0.00                   0.979592   
5393889       0.555556                0.25                   0.795918   
5393887       0.629630                0.25                   0.153061   
5393863       0.407407                0.75                   0.306122   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                    0.100000                0.588889  0.391192   
5393091                    0.533333                0.322222  0.457898   
5393889                    0.066667                0.011111  0.305696   
5393887                    0.466667                0.300000  0.306989   
5393863                    0.100000                0.322222  0.234815   

         Agreement Reached  Number of Dependents  
5393875                0.0              0.166667  
5393091                1.0              0.666667  
5393889                0.0              1.000000  
5393887                0.0              0.166667  
5393863                0.0              0.833333

In [ ]:
#df_scaled.set_index(df_scaled.columns[0], inplace=True)

In [62]:
df_scaled_imputed = df_scaled.copy()

In [63]:
imp = IterativeImputer(
    estimator=RandomForestRegressor(n_jobs=-1, random_state=17)
    ,initial_strategy='median'
    ,max_iter=5
    ,random_state=17
    ,verbose=2
)

In [64]:
fit_sample = df_scaled_imputed.sample(n=30000, random_state=17)

In [65]:
imp.fit(fit_sample)

[IterativeImputer] Completing matrix with shape (30000, 26)
[IterativeImputer] Ending imputation round 1/5, elapsed time 59.71
[IterativeImputer] Change: 1.5547486545526927, scaled tolerance: 0.001 
[IterativeImputer] Ending imputation round 2/5, elapsed time 128.44
[IterativeImputer] Change: 1.0361577833739262, scaled tolerance: 0.001 
[IterativeImputer] Ending imputation round 3/5, elapsed time 198.44
[IterativeImputer] Change: 0.836141464142012, scaled tolerance: 0.001 
[IterativeImputer] Ending imputation round 4/5, elapsed time 266.99
[IterativeImputer] Change: 0.8454298441470212, scaled tolerance: 0.001 
[IterativeImputer] Ending imputation round 5/5, elapsed time 334.87
[IterativeImputer] Change: 0.699772085033756, scaled tolerance: 0.001 


C:\Users\joaom\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\impute\_iterative.py:825: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


IterativeImputer(estimator=RandomForestRegressor(n_jobs=-1, random_state=17),
                 initial_strategy='median', max_iter=5, random_state=17,
                 verbose=2)

In [66]:
df_scaled_imputed = imp.transform(df_scaled_imputed)

[IterativeImputer] Completing matrix with shape (559202, 26)
[IterativeImputer] Ending imputation round 1/5, elapsed time 3.90
[IterativeImputer] Ending imputation round 2/5, elapsed time 7.27
[IterativeImputer] Ending imputation round 3/5, elapsed time 11.46
[IterativeImputer] Ending imputation round 4/5, elapsed time 14.72
[IterativeImputer] Ending imputation round 5/5, elapsed time 17.15


In [67]:
df_scaled_imputed = pd.DataFrame(df_scaled_imputed, columns = df_scaled.columns).set_index(df_scaled.index)

In [68]:
df_scaled_imputed.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875       0.939606       0.258333                             0.0   
5393091       0.934224       0.383333                             0.0   
5393889       0.938548       0.333333                             0.0   
5393887       0.939606       0.508333                             0.0   
5393863       0.939430       0.558333                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875            0.0                      0.0             0.000000   
5393091            0.0                      1.0             0.062790   
5393889            0.0                      0.0             0.051601   
5393887            0.0                      0.0             0.000000   
5393863            0.0                      0.0             0.000000   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875    0.985134  0.848420  0.195237      0.585330      0.000000   
5393091    0.977701  0.848517  0.184600      0.999511      0.000000   
5393889    0.980674  0.848517  0.195558      0.437164      0.000000   
5393887    0.970268  0.848420  0.187471      0.836186      0.142857   
5393863    0.967294  0.848420  0.190989      0.436675      0.000000   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875           0.142857          0.790323                 0.0   
5393091           0.428571          0.983871                 0.0   
5393889           0.428571          0.564516                 0.0   
5393887           0.142857          0.209677                 0.0   
5393863           0.285714          0.822581                 0.0   

         District Name  First Hearing Date    Gender  IME-4 Count  \
5393875       1.000000            0.066132  0.333333     0.032639   
5393091       0.714286            0.013836  0.000000     0.041667   
5393889       0.000000            0.181535  0.333333     0.027222   
5393887       0.000000            0.083623  0.333333     0.024444   
5393863       0.428571            0.141535  0.333333     0.026806   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875       0.407407                0.00                   0.265306   
5393091       0.148148                0.00                   0.979592   
5393889       0.555556                0.25                   0.795918   
5393887       0.629630                0.25                   0.153061   
5393863       0.407407                0.75                   0.306122   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                    0.100000                0.588889  0.391192   
5393091                    0.533333                0.322222  0.457898   
5393889                    0.066667                0.011111  0.305696   
5393887                    0.466667                0.300000  0.306989   
5393863                    0.100000                0.322222  0.234815   

         Agreement Reached  Number of Dependents  
5393875                0.0              0.166667  
5393091                1.0              0.666667  
5393889                0.0              1.000000  
5393887                0.0              0.166667  
5393863                0.0              0.833333

In [69]:
df_scaled_imputed.isna().sum()

Accident Date                     0
Age at Injury                     0
Alternative Dispute Resolution    0
Assembly Date                     0
Attorney/Representative           0
Average Weekly Wage               0
Birth Year                        0
C-2 Date                          0
C-3 Date                          0
Carrier Name                      0
Carrier Type                      0
Claim Injury Type                 0
County of Injury                  0
COVID-19 Indicator                0
District Name                     0
First Hearing Date                0
Gender                            0
IME-4 Count                       0
Industry Code                     0
Medical Fee Region                0
WCIO Cause of Injury Code         0
WCIO Nature of Injury Code        0
WCIO Part Of Body Code            0
Zip Code                          0
Agreement Reached                 0
Number of Dependents              0
dtype: int64

In [70]:
df_scaled.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875       0.939606       0.258333                             0.0   
5393091       0.934224       0.383333                             0.0   
5393889       0.938548       0.333333                             0.0   
5393887       0.939606       0.508333                             0.0   
5393863       0.939430       0.558333                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875            0.0                      0.0             0.000000   
5393091            0.0                      1.0             0.062790   
5393889            0.0                      0.0             0.051601   
5393887            0.0                      0.0                  NaN   
5393863            0.0                      0.0             0.000000   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875    0.985134  0.848420       NaN      0.585330      0.000000   
5393091    0.977701  0.848517    0.1846      0.999511      0.000000   
5393889    0.980674  0.848517       NaN      0.437164      0.000000   
5393887    0.970268  0.848420       NaN      0.836186      0.142857   
5393863    0.967294  0.848420       NaN      0.436675      0.000000   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875           0.142857          0.790323                 0.0   
5393091           0.428571          0.983871                 0.0   
5393889           0.428571          0.564516                 0.0   
5393887           0.142857          0.209677                 0.0   
5393863           0.285714          0.822581                 0.0   

         District Name  First Hearing Date    Gender  IME-4 Count  \
5393875       1.000000                 NaN  0.333333          NaN   
5393091       0.714286            0.013836  0.000000     0.041667   
5393889       0.000000                 NaN  0.333333          NaN   
5393887       0.000000                 NaN  0.333333          NaN   
5393863       0.428571                 NaN  0.333333          NaN   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875       0.407407                0.00                   0.265306   
5393091       0.148148                0.00                   0.979592   
5393889       0.555556                0.25                   0.795918   
5393887       0.629630                0.25                   0.153061   
5393863       0.407407                0.75                   0.306122   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                    0.100000                0.588889  0.391192   
5393091                    0.533333                0.322222  0.457898   
5393889                    0.066667                0.011111  0.305696   
5393887                    0.466667                0.300000  0.306989   
5393863                    0.100000                0.322222  0.234815   

         Agreement Reached  Number of Dependents  
5393875                0.0              0.166667  
5393091                1.0              0.666667  
5393889                0.0              1.000000  
5393887                0.0              0.166667  
5393863                0.0              0.833333

In [71]:
df_scaled[missing_values] = df_scaled_imputed[missing_values]

In [72]:
df_scaled.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
5393875       0.939606       0.258333                             0.0   
5393091       0.934224       0.383333                             0.0   
5393889       0.938548       0.333333                             0.0   
5393887       0.939606       0.508333                             0.0   
5393863       0.939430       0.558333                             0.0   

         Assembly Date  Attorney/Representative  Average Weekly Wage  \
5393875            0.0                      0.0             0.000000   
5393091            0.0                      1.0             0.062790   
5393889            0.0                      0.0             0.051601   
5393887            0.0                      0.0             0.000000   
5393863            0.0                      0.0             0.000000   

         Birth Year  C-2 Date  C-3 Date  Carrier Name  Carrier Type  \
5393875    0.985134  0.848420       NaN      0.585330      0.000000   
5393091    0.977701  0.848517    0.1846      0.999511      0.000000   
5393889    0.980674  0.848517       NaN      0.437164      0.000000   
5393887    0.970268  0.848420       NaN      0.836186      0.142857   
5393863    0.967294  0.848420       NaN      0.436675      0.000000   

         Claim Injury Type  County of Injury  COVID-19 Indicator  \
5393875           0.142857          0.790323                 0.0   
5393091           0.428571          0.983871                 0.0   
5393889           0.428571          0.564516                 0.0   
5393887           0.142857          0.209677                 0.0   
5393863           0.285714          0.822581                 0.0   

         District Name  First Hearing Date    Gender  IME-4 Count  \
5393875       1.000000                 NaN  0.333333          NaN   
5393091       0.714286            0.013836  0.000000     0.041667   
5393889       0.000000                 NaN  0.333333          NaN   
5393887       0.000000                 NaN  0.333333          NaN   
5393863       0.428571                 NaN  0.333333          NaN   

         Industry Code  Medical Fee Region  WCIO Cause of Injury Code  \
5393875       0.407407                0.00                   0.265306   
5393091       0.148148                0.00                   0.979592   
5393889       0.555556                0.25                   0.795918   
5393887       0.629630                0.25                   0.153061   
5393863       0.407407                0.75                   0.306122   

         WCIO Nature of Injury Code  WCIO Part Of Body Code  Zip Code  \
5393875                    0.100000                0.588889  0.391192   
5393091                    0.533333                0.322222  0.457898   
5393889                    0.066667                0.011111  0.305696   
5393887                    0.466667                0.300000  0.306989   
5393863                    0.100000                0.322222  0.234815   

         Agreement Reached  Number of Dependents  
5393875                0.0              0.166667  
5393091                1.0              0.666667  
5393889                0.0              1.000000  
5393887                0.0              0.166667  
5393863                0.0              0.833333

In [73]:
df_scaled.isna().sum()

Accident Date                          0
Age at Injury                          0
Alternative Dispute Resolution         0
Assembly Date                          0
Attorney/Representative                0
Average Weekly Wage                    0
Birth Year                             0
C-2 Date                           12746
C-3 Date                          378915
Carrier Name                           0
Carrier Type                           0
Claim Injury Type                      0
County of Injury                       0
COVID-19 Indicator                     0
District Name                          0
First Hearing Date                414888
Gender                                 0
IME-4 Count                       430115
Industry Code                          0
Medical Fee Region                     0
WCIO Cause of Injury Code              0
WCIO Nature of Injury Code             0
WCIO Part Of Body Code                 0
Zip Code                               0
Agreement Reache

In [ ]:
#df_scaled.to_csv('/home/shadybea/OneDrive/General/Machine Learning/Project/Data/train_data_scaled_imputed.csv', index=True)

In [ ]:
#df = pd.read_csv('/home/shadybea/OneDrive/General/Machine Learning/Project/Data/train_data_scaled_imputed.csv', sep=',')

In [ ]:
#df.set_index(df.columns[0], inplace=True)

In [74]:
def mode(x): return x.mode().iloc[0] if not x.mode().empty else None


In [75]:
df['C-3 Date Missing'] = df['C-3 Date'].isna().map({True: 1, False: 0})

In [76]:
df[['C-3 Date Missing', 'Claim Injury Type']].groupby(['Claim Injury Type', 'C-3 Date Missing'])['C-3 Date Missing'].agg(
    frequency=lambda x: x.value_counts(),
    proportion=lambda x: x.value_counts() / len(df) * 100,
    mode=mode
)

frequency  proportion  mode
Claim Injury Type C-3 Date Missing                             
1. CANCELLED      0                      5609    0.977135     0
                  1                      6867    1.196289     1
2. NON-COMP       0                     38127    6.642045     0
                  1                    252951   44.066199     1
3. MED ONLY       0                     23841    4.153303     0
                  1                     45065    7.850703     1
4. TEMPORARY      0                     78974   13.757937     0
                  1                     69533   12.113235     1
5. PPD SCH LOSS   0                     37215    6.483167     0
                  1                     11065    1.927616     1
6. PPD NSL        0                      3352    0.583947     0
                  1                       859    0.149645     1
7. PTD            0                        79    0.013762     0
                  1                        18    0.003136     1
8. DEATH          0                        47    0.008188     0
                  1                       423    0.073690     1